### PR TITLE
Add Ability to Enable/Disable Shortcuts and Selection for LineEdit/TextEdit

### LINKED ISSUE #31183: Text Selection Forced when using LineEdit/TextEdit
Godot 3.1.1

I am interested in implementing a custom text input field.
To do this, I have 2 options:
1. Use the functionality of LineEdit TextEdit and hide the styling
2. Read input manually

If I go with LineEdit, I can update the text, however I am forced to support visuals for selections. Because [It features many built-in shortcuts which will always be available](https://docs.godotengine.org/en/3.1/classes/class_lineedit.html#description). 

In other words, I have no visuals for text selection and do not want to support it. However disabling this feature is not possible using LineEdit/TextEdit nodes.

Going with the other option gimps me in terms of supporting unicode characters, decoding which from scancode/unicode symbols is not doable (ąčęėįš codes do not convert into the appropriate string letters using ` PoolByteArray([event.unicode]).get_string_from_utf8()` magic from [here](https://github.com/godotengine/godot/issues/6166)).

**TL;DR: please allow disabling text selection and/or shortcuts available when editing text.**

### PR DESCRIPTION
**Pull request by codecustard**
This PR adds the ability to enable/disable shortcut keys and selection for LineEdit and TextEdit.

Closes: #31183

### CODE REVIEW

File: scene/gui/line_edit.cpp [hunk 1]
```diff
@@ -203,29 +203,30 @@ void LineEdit::_gui_input(Ref<InputEvent> p_event) {
 
 				case (KEY_X): { // CUT
 
-					if (editable) {
+					if (editable && is_shortcut_keys_enabled()) {
```
Comment (akien-mga) [inline scene/gui/line_edit.cpp]: Instead of adding this check in all cases, maybe simply put it in the outer `if`?

File: scene/gui/line_edit.cpp [hunk 2]
```diff
@@ -1524,6 +1531,30 @@ bool LineEdit::is_clear_button_enabled() const {
 	return clear_button_enabled;
 }
 
+void LineEdit::set_shortcut_keys_enabled(bool p_enabled) {
+	shortcut_keys_enabled = p_enabled;
+
+	// Reorganize context menu.
```
Comment (akien-mga) [inline scene/gui/line_edit.cpp]: Might be worth moving to a private method since it's called in two places.

**Comment by akien-mga** [comment]
> Closes: #31183

This issue also requests the same feature in TextEdit.

**Comment by akien-mga** [comment]
Needs a rebase to remove the merge commit and fix merge conflicts, otherwise I guess it should be ok to merge.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
This doesn't seem to fix the linked issue? As it describes disabling selections not shortcuts.

In any case, you should only need to remove the shortcuts inside new method `_generate_context_menu` leaving the main switch alone.

**Comment by sparkart** [comment]
> This doesn't seem to fix the linked issue? As it describes disabling selections not shortcuts.
> 
> In any case, you should only need to remove the shortcuts inside new method `_generate_context_menu` leaving the main switch alone.

I've tried that before, but it does not seem to have any effect. The context menu will show or not show the shortcut keys... but the switch statement causes the shortcut keys to have effect regardless of whether the context menu has them or not.

**Comment by Paulb23** [comment]
> I've tried that before, but it does not seem to have any effect

Ahh my bad, I remember now, currently they are visual only. Related #29490, which will require shortcuts to handle echoed events before they can be changed.

In that case, there are a couple missing:

* `select_all()` on line 3459 
* `cut()` on line 3201

File: scene/gui/line_edit.cpp [hunk 3]
```diff
@@ -1533,6 +1512,26 @@ bool LineEdit::is_clear_button_enabled() const {
...
+}
+
+bool LineEdit::is_shortcut_keys_enabled() const {
+	return shortcut_keys_enabled;
+}
+
+void LineEdit::set_selecting_enabled(bool p_enabled) {
+	selecting_enabled = p_enabled;
+
+	_generate_context_menu();
```
Comment (Paulb23) [inline scene/gui/line_edit.cpp]: If this is set to false, it should clear ant current selection.

File: scene/gui/text_edit.cpp [hunk 4]
```diff
@@ -6764,6 +6799,26 @@ bool TextEdit::is_context_menu_enabled() {
...
+void TextEdit::set_shortcut_keys_enabled(bool p_enabled) {
+	shortcut_keys_enabled = p_enabled;
+
+	_generate_context_menu();
+}
+
+void TextEdit::set_selecting_enabled(bool p_enabled) {
+	selecting_enabled = p_enabled;
+
+	_generate_context_menu();
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: Same here.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Couple notes left inline, It's also still possible to select with the mouse or shift + arrow keys within `line_edit`.

**Review by Paulb23 — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**